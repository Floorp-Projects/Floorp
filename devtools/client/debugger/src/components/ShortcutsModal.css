/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.shortcuts-content {
  padding: 15px;
  column-width: 250px;
  cursor: default;
  user-select: none;
}

.shortcuts-content h2 {
  margin-top: 2px;
  margin-bottom: 2px;
  color: var(--theme-text-color-strong);
}

.shortcuts-section {
  display: inline-block;
  margin: 5px;
  margin-bottom: 15px;
  width: 250px;
}

.shortcuts-list {
  list-style: none;
  margin: 0px;
  padding: 0px;
  overflow: auto;
  width: calc(100% - 1px); /* 1px fixes the hidden right border */
}

.shortcuts-list li {
  font-size: 12px;
  color: var(--theme-body-color);
  padding-top: 5px;
  display: flex;
  justify-content: space-between;
  border: 1px solid transparent;
  white-space: pre;
}

@media (max-width: 640px) {
  .shortcuts-section {
    width: 100%;
  }
}
