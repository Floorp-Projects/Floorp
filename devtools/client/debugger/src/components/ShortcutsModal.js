/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import Modal from "./shared/Modal";
import classnames from "classnames";
import { formatKeyShortcut } from "../utils/text";

import "./ShortcutsModal.css";

export class ShortcutsModal extends Component {
  renderPrettyCombos(combo) {
    return combo
      .split(" ")
      .map(c => (
        <span key={c} className="keystroke">
          {c}
        </span>
      ))
      .reduce((prev, curr) => [prev, " + ", curr]);
  }

  renderShorcutItem(title, combo) {
    return (
      <li>
        <span>{title}</span>
        <span>{this.renderPrettyCombos(combo)}</span>
      </li>
    );
  }

  renderEditorShortcuts() {
    return (
      <ul className="shortcuts-list">
        {this.renderShorcutItem(
          L10N.getStr("shortcuts.toggleBreakpoint"),
          formatKeyShortcut(L10N.getStr("toggleBreakpoint.key"))
        )}
        {this.renderShorcutItem(
          L10N.getStr("shortcuts.toggleCondPanel.breakpoint"),
          formatKeyShortcut(L10N.getStr("toggleCondPanel.breakpoint.key"))
        )}
        {this.renderShorcutItem(
          L10N.getStr("shortcuts.toggleCondPanel.logPoint"),
          formatKeyShortcut(L10N.getStr("toggleCondPanel.logPoint.key"))
        )}
      </ul>
    );
  }

  renderSteppingShortcuts() {
    return (
      <ul className="shortcuts-list">
        {this.renderShorcutItem(L10N.getStr("shortcuts.pauseOrResume"), "F8")}
        {this.renderShorcutItem(L10N.getStr("shortcuts.stepOver"), "F10")}
        {this.renderShorcutItem(L10N.getStr("shortcuts.stepIn"), "F11")}
        {this.renderShorcutItem(
          L10N.getStr("shortcuts.stepOut"),
          formatKeyShortcut(L10N.getStr("stepOut.key"))
        )}
      </ul>
    );
  }

  renderSearchShortcuts() {
    return (
      <ul className="shortcuts-list">
        {this.renderShorcutItem(
          L10N.getStr("shortcuts.fileSearch2"),
          formatKeyShortcut(L10N.getStr("sources.search.key2"))
        )}
        {this.renderShorcutItem(
          L10N.getStr("shortcuts.searchAgain2"),
          formatKeyShortcut(L10N.getStr("sourceSearch.search.again.key3"))
        )}
        {this.renderShorcutItem(
          L10N.getStr("shortcuts.projectSearch2"),
          formatKeyShortcut(L10N.getStr("projectTextSearch.key"))
        )}
        {this.renderShorcutItem(
          L10N.getStr("shortcuts.functionSearch2"),
          formatKeyShortcut(L10N.getStr("functionSearch.key"))
        )}
        {this.renderShorcutItem(
          L10N.getStr("shortcuts.gotoLine"),
          formatKeyShortcut(L10N.getStr("gotoLineModal.key3"))
        )}
      </ul>
    );
  }

  renderShortcutsContent() {
    return (
      <div
        className={classnames("shortcuts-content", this.props.additionalClass)}
      >
        <div className="shortcuts-section">
          <h2>{L10N.getStr("shortcuts.header.editor")}</h2>
          {this.renderEditorShortcuts()}
        </div>
        <div className="shortcuts-section">
          <h2>{L10N.getStr("shortcuts.header.stepping")}</h2>
          {this.renderSteppingShortcuts()}
        </div>
        <div className="shortcuts-section">
          <h2>{L10N.getStr("shortcuts.header.search")}</h2>
          {this.renderSearchShortcuts()}
        </div>
      </div>
    );
  }

  render() {
    const { enabled } = this.props;

    if (!enabled) {
      return null;
    }

    return (
      <Modal
        in={enabled}
        additionalClass="shortcuts-modal"
        handleClose={this.props.handleClose}
      >
        {this.renderShortcutsContent()}
      </Modal>
    );
  }
}
