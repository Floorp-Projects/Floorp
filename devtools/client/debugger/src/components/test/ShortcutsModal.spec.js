/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";
import { ShortcutsModal } from "../ShortcutsModal";

function render(overrides = {}) {
  const props = {
    enabled: true,
    handleClose: jest.fn(),
    additionalClass: "",
    ...overrides,
  };
  const component = shallow(<ShortcutsModal {...props} />);

  return { component, props };
}

describe("ShortcutsModal", () => {
  it("renders when enabled", () => {
    const { component } = render();
    expect(component).toMatchSnapshot();
  });

  it("renders nothing when not enabled", () => {
    const { component } = render({
      enabled: false,
    });
    expect(component.text()).toBe("");
  });

  it("renders with additional classname", () => {
    const { component } = render({
      additionalClass: "additional-class",
    });
    expect(
      component.find(".shortcuts-content").hasClass("additional-class")
    ).toEqual(true);
  });
});
