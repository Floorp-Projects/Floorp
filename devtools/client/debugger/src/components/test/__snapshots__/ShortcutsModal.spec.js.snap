// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShortcutsModal renders when enabled 1`] = `
<Slide
  additionalClass="shortcuts-modal"
  handleClose={[MockFunction]}
  in={true}
>
  <div
    className="shortcuts-content"
  >
    <div
      className="shortcuts-section"
    >
      <h2>
        Editor
      </h2>
      <ul
        className="shortcuts-list"
      >
        <li>
          <span>
            Toggle Breakpoint
          </span>
          <span>
            <span
              className="keystroke"
              key="Ctrl+B"
            >
              Ctrl+B
            </span>
          </span>
        </li>
        <li>
          <span>
            Edit Conditional Breakpoint
          </span>
          <span>
            <span
              className="keystroke"
              key="Ctrl+Shift+B"
            >
              Ctrl+Shift+B
            </span>
          </span>
        </li>
        <li>
          <span>
            Edit Log Point
          </span>
          <span>
            <span
              className="keystroke"
              key="Ctrl+Shift+Y"
            >
              Ctrl+Shift+Y
            </span>
          </span>
        </li>
      </ul>
    </div>
    <div
      className="shortcuts-section"
    >
      <h2>
        Stepping
      </h2>
      <ul
        className="shortcuts-list"
      >
        <li>
          <span>
            Pause/Resume
          </span>
          <span>
            <span
              className="keystroke"
              key="F8"
            >
              F8
            </span>
          </span>
        </li>
        <li>
          <span>
            Step Over
          </span>
          <span>
            <span
              className="keystroke"
              key="F10"
            >
              F10
            </span>
          </span>
        </li>
        <li>
          <span>
            Step In
          </span>
          <span>
            <span
              className="keystroke"
              key="F11"
            >
              F11
            </span>
          </span>
        </li>
        <li>
          <span>
            Step Out
          </span>
          <span>
            <span
              className="keystroke"
              key="Shift+F11"
            >
              Shift+F11
            </span>
          </span>
        </li>
      </ul>
    </div>
    <div
      className="shortcuts-section"
    >
      <h2>
        Search
      </h2>
      <ul
        className="shortcuts-list"
      >
        <li>
          <span>
            Go to file
          </span>
          <span>
            <span
              className="keystroke"
              key="Ctrl+P"
            >
              Ctrl+P
            </span>
          </span>
        </li>
        <li>
          <span>
            Find in files
          </span>
          <span>
            <span
              className="keystroke"
              key="Ctrl+Shift+F"
            >
              Ctrl+Shift+F
            </span>
          </span>
        </li>
        <li>
          <span>
            Find function
          </span>
          <span>
            <span
              className="keystroke"
              key="Ctrl+Shift+O"
            >
              Ctrl+Shift+O
            </span>
          </span>
        </li>
        <li>
          <span>
            Go to line
          </span>
          <span>
            <span
              className="keystroke"
              key="Ctrl+G"
            >
              Ctrl+G
            </span>
          </span>
        </li>
      </ul>
    </div>
  </div>
</Slide>
`;
